### ISSUE TITLE
We need a solid strategy for dealing with binary dependencies in catapult project.

### ISSUE THREAD

**Issue by nedn** [catapult-project/catapult#1810]
There are various projects in catapult that require binary files as dependencies, e.g:
vinn/ requires d8 binary to run js.
telemetry/ requires adb for communicating with android devices, bitmaptools,..

In the past, we have been dealing with these issues in hacky ways:
1) d8 binary is checked in vinn/third_party/ directly.
2) telemetry download the required binary from cloud storage when it needed.

Those way of dealing with binary dependencies are far from ideal because:
+ For 1), git is not designed to deal with binary files well. Checking in binaries in git blows up the git commit tree, and slow down git operations significantly.
+ For 2), downloading binaries from cloud storage during runtime leads to flaky failures, make the test operation less deterministic (non idempotent when running under isolated testing infra of Chromium), and causes racy issue of downloading binaries when multiple tests are run in parallel (see crbug.com/559154).

Here are a few solutions that have been discussed:
1) Use @Apeliotes 's dependencies manager for fetching all the binaries deps in one pre step ("gclient runhooks", or setup.py). 
2) Do binary releases (@maruel 's idea), assuming outside customers would want to use the "releases" instead of "git clone".  

fyi: @natduca @anniesullie @vinterstum @simonhatch

**Comment by anniesullie** [comment]
Another use case for having a solid strategy: https://docs.google.com/document/d/1PZmyX_hmeXiUXdXgwVcw0kHMSjlnmtTAnp4VnBw8Gtk/edit

One thing I really like about 2) is that it's mostly invisible to the user, and we don't need to teach them new commands. I thought the direction we were going for swarming/isolates was to have the binary manager be able to do a single pre step, but also download on-demand if that pre-step didn't run. Is that too complex?

**Comment by nedn** [comment]
@anniesullie if we download all the binary in a single pre step, then we should disable download on-demand entirely to make the tests idempotent (https://github.com/luci/luci-py/blob/0e6083e78c58cacb3040a65df6a7924200c4d5a5/appengine/swarming/doc/User-Guide.md#task-idempotency).

**Comment by anniesullie** [comment]
@nedn: Couldn't we just have a flag that we use on the buildbots to disable on-demand download entirely, but still allow end users to do that so they don't need to remember to run the setup script?

**Comment by nedn** [comment]
If we worry about end-users forget to run the setup scripts, there are few other options:
1) We ask the end users to use binary releases as mentioned above. No download is needed since the releases tarball should have everything packaged.
2) Telemetry 's run_benchmarks or other scripts in catapult will download all the required binaries if they aren't up-to-date as the first step.
3) Telemetry 's run_benchmarks or other scripts in catapult will blow up with an error message "Need to run setup.py" first if there are any binaries update.

If we go to the download route for end user, I would prefer downloading all the binaries in one run, and blow up if any of them fail rather than downloading individual ones which may make the program runtime environment be in unexpected state (e.g: binary X, Y are downloaded, binary Z is incomplete..)

**Comment by anniesullie** [comment]
Yeah, having them downloaded up front automatically for end users would be great. @Apeliotes how hard is this for the binary manager to do?

**Comment by maruel** [comment]
I'm feeling a bit awkward to state my opinion because I'm neither on the team neither a target user but here it is:

> One thing I really like about 2) is that it's mostly invisible to the user

I disagree with this assessment. It's mostly a black box to the user and when things go wrong, it's impossible to debug. Having the step explicit (either prepackaging via a release or as a specific documentated preparation step) removes the black box halo.

What about corp users behind a non-transparent HTTP proxy? Good luck finding the place where you want to type your password.

Not having arbitrary remote request during an otherwise _100% local operation_ is better for both user diagnosis and user privacy reasons. I'd be mad if while on a flight I wouldn't be able to run my stuff I had prepared before departure. In general my opinion boils down to "_Do not make a 100% local operation non-local by a non-default unexpected side effect_".

**Comment by natduca** [comment]
I am sympathetic to the binary distribution approach as a long term goal, actually. But, I think there's a sequencing argument to be made here s.t. option 3 is the starting point and then we revisit binary releases down the road.

The big picture, from my PoV, is that we want to make Catapult be a A Stable Thing. A project you can get ahold of and provides you with the stuff to go Fast. We're not there yet because we're still bootstrapping it -- namely, we want to get telemetry moved over into catapult. That's p1.

The binary distribution solves many problems in the "right way" but transitioning to it brings a lot of workflow complications to the project during the transition stage. The obvious way to deal with those complications would be either to have our eng teams deal with the bumps, or just slow down the transition until we get all the binary distription crinkles sorted out.

Telemetry->catapult remains highly important, so I think I'd prefer something more expedient. Then, as we all carry along with the catapult journey, we keep an eye on issues as the crop up and if we start seeing the right critical mass, we come back to the binary distro option and consider again whether its time.

**Comment by vinterstum** [comment]
One random data point:

We ran into an issue the other day, where the vinn d8 binary is built
against a specific distro of Linux (i guess current Ubuntu version), which
means having a dependency on a never glibc version than the App Engine
Managed VMs run with. It'd be nice if there was less friction here, as this
will also apply to external users.

No idea what the right solution is however. I'm guessing we don't want to
build C++ targets in Catapult, but without that we'll need to choose a
specific set of supported platforms.

**Comment by nedn** [comment]
At this point, I think option (3), i.e: providing a install-catapult-build-deps script and having all of our scripts prompt users to run the script if the binaries are not downloaded/or up-to-date is the best option for us to move forward.

@Apeliotes 's dependency manager system (currently in telemetry/) does know how to handle versioning of dependencies, so it's ideal if we block this bug on telemetry/ move to catapult.

**Comment by natduca** [comment]
Is this resolved?

**Comment by nedn** [comment]
@Apeliotes Can you drive this?

**Comment by Apeliotes** [comment]
Something that hasn't been talked about here yet: where do we want these dependencies to be downloaded? 
My suggestion is third_party/bin for catapult dependencies. Project specific dependencies may have a different solution (especially if they need a different version of a dependency than other catapult projects.)

Also note that I'm in the process of creating a catapult-wide solution for stable chrome in [this bug](https://github.com/catapult-project/catapult/issues/1968). I'm approaching the implementation in a way that will help simplify fetching all of the binaries from a setup script.
